engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala

```diff
@@ -27,39 +25,64 @@ class TypeMethodReference(methodReference: MethodReference, currentResults: List
         Right(Unknown)
     }
 
-  private lazy val paramsCount = methodReference.getChildCount
-
   private def extractClazzDefinitions(typedClasses: Set[SingleTypingResult])(implicit settings: ClassExtractionSettings): List[ClazzDefinition] =
     typedClasses.map(typedClass =>
       EspTypeUtils.clazzDefinition(typedClass.objType.klass)
     ).toList
 
-  private def typeFromClazzDefinitions(clazzDefinitions: List[ClazzDefinition]): Either[String, TypingResult] =
-    clazzDefinitions match {
-      case Nil =>
-        Right(Unknown)
-      case _ =>
-        val isClass = clazzDefinitions.map(k => k.clazzName).exists(_.canBeSubclassOf(Typed[Class[_]]))
-        val display = clazzDefinitions.map(k => k.clazzName).map(_.display).mkString(", ")
-        clazzDefinitions.flatMap(_.methods.get(methodReference.getName)) match {
-          //Static method can be invoked - we cannot find them ATM
-          case Nil if isClass => Right(Unknown)
-          case Nil  => Left(s"Unknown method '${methodReference.getName}' in $display")
-          case methodInfoes => typeFromMethodInfoes(methodInfoes)
-        }
+  private def typeFromClazzDefinitions(clazzDefinitions: List[ClazzDefinition]): Either[String, TypingResult] = {
+    val validatedType = for {
+      nonEmptyClassDefinitions <- validateClassDefinitionsNonEmpty(clazzDefinitions).right
+      nonEmptyMethods <- validateMethodsNonEmpty(nonEmptyClassDefinitions).right
+      methodsWithMatchingArity <- validateMethodArity(nonEmptyMethods).right
+      returnTypesForMatchingParams <- validateMethodParameterTypes(methodsWithMatchingArity).right
+    } yield Typed(returnTypesForMatchingParams.toSet)
+
+    validatedType match {
+      case Left(None) => Right(Unknown) // we use Left(None) to indicate situation when we want to skip further validations because of lack of some knowledge
+      case Left(Some(message)) => Left(message)
+      case Right(returnType) => Right(returnType)
+    }
+  }
+
+  private def validateClassDefinitionsNonEmpty(clazzDefinitions: List[ClazzDefinition]): Either[Option[String], List[ClazzDefinition]] =
+    if (clazzDefinitions.isEmpty) Left(None) else Right(clazzDefinitions)
```

Comment:
does empty list represent Unknown here?

Reply:
Yep, I've described this monadic rule only in comment here `case Left(None)`. Is it enough or you think that we should add more descriptive type for left part?

Reply:
No, it's ok, maybe a comment would be helpful